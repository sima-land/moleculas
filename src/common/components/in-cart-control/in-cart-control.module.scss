@use 'node_modules/@sima-land/ui-nucleons/colors';

.control {
  flex-shrink: 0;
  width: 122px;
}

.button-spinner {
  fill: none;
  animation: rotation 3s linear infinite;
}

@keyframes rotation {
  to {
    transform: rotate(360deg);
  }
}
