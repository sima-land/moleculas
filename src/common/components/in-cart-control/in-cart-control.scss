@use 'node_modules/@dev-dep/ui-nucleons/colors';

.control {
  flex-shrink: 0;
  width: 122px;
}

.button-spinner {
  fill: none;
  animation: rotation 3s linear infinite;
}

@keyframes rotation {
  to {
    transform: rotate(360deg);
  }
}
