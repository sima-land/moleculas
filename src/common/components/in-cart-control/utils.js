import { useEffect, useState } from 'react';
import { formatQtyString } from '../../helpers/format-qty-string';

/**
 * Возвращает свойства input, которые фиксируют поведение управления количеством в корзине.
 * @param {{ qty, allowFloat, onApply }} props Свойства.
 * @return {{ value, onChange, onKeyDown, onBlur }} Свойства input.
 */
export const useCartItemInput = ({ qty: qtyArg, allowFloat, onApply }) => {
  const qty = Number.isFinite(qtyArg) ? qtyArg : 0;

  const [value, setValue] = useState(String(qty));

  // eslint-disable-next-line require-jsdoc
  const checkNeedApply = () => {
    const parsedQty = parseFloat(value) || 0; // ноль, так как можно ввести пустую сроку
    const isDifferent = parsedQty !== qty;

    isDifferent && onApply && onApply(parsedQty);
  };

  useEffect(() => setValue(String(qty)), [qty]);

  return {
    value,

    onChange: e => {
      setValue(formatQtyString(e.target.value, { allowFloat }));
    },

    onKeyDown: e => {
      e.key === 'Enter' && checkNeedApply();
    },

    onBlur: checkNeedApply,
  };
};
