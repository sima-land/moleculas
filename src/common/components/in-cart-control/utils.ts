import { useCallback, useEffect, useRef, useState } from 'react';
import { formatQtyString } from '../../helpers/format-qty-string';
import { useIdentityRef } from '@sima-land/ui-nucleons/hooks/identity';

/**
 * Возвращает свойства input, которые фиксируют поведение управления количеством в корзине.
 * @param props Свойства.
 * @return Свойства input.
 */
export const useCartItemInput = ({ qty: qtyArg, allowFloat, onApply }: {
  qty: number
  onApply: (newQty: number) => void
  allowFloat?: boolean
}) => {
  const qty = Number.isFinite(qtyArg) ? qtyArg : 0;

  const [value, setValue] = useState(String(qty));

  const focused = useRef(false);

  const checker = useIdentityRef(useCallback(() => {
    const parsedQty = parseFloat(value) || 0; // ноль, так как можно ввести пустую сроку
    const isDifferent = parsedQty !== qty;

    isDifferent && onApply(parsedQty);
  }, [qty, value, onApply]));

  useEffect(() => setValue(String(qty)), [qty]);

  // blur не происходит при размонтировании - учитываем
  useEffect(() => () => {
    focused.current && checker.current();
  }, []);

  return {
    value,

    onChange: useCallback(e => {
      setValue(formatQtyString(e.target.value, { allowFloat }));
    }, [allowFloat]),

    onKeyDown: useCallback(e => {
      e.key === 'Enter' && checker.current();
    }, []),

    onFocus: useCallback(() => {
      focused.current = true;
    }, []),

    onBlur: useCallback(() => {
      focused.current = false;
      checker.current();
    }, []),
  };
};
