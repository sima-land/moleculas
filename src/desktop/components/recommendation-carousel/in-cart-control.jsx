import React from 'react';
import { Button } from '@dev-dep/ui-nucleons/button';
import { Stepper } from '@dev-dep/ui-nucleons/stepper';
import classnames from 'classnames/bind';
import styles from './in-cart-control.scss';
import { noop } from 'lodash';

const cx = classnames.bind(styles);

/**
 * Заглушка кнопки/степпера управления количеством в корзине.
 * @param {Object} props Свойства.
 * @return {ReactElement} Элемент.
 */
export const InCartControl = ({
  canAdd,
  canSubtract,
  className,
  onAdd,
  onChange,
  onSubtract,
  qty,
}) => (
  <>
    {
      qty > 0
        ? (
          <Stepper
            size='s'
            className={cx('control', className)}
            value={qty}
            canAdd={canAdd}
            canSubtract={canSubtract}
            onAdd={onAdd}
            onSubtract={onSubtract}
            onKeyDown={e => e.key === 'Enter' && onChange(e)}
            onBlur={onChange}
            onChange={noop}
          />
        ) : (
          <Button
            size='small'
            className={cx('control', className)}
            onClick={onAdd}
          >
            В корзину
          </Button>
        )
    }
  </>
);
