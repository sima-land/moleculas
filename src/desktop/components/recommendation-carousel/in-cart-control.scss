@use 'node_modules/@dev-dep/ui-nucleons/colors';

.control {
  flex-shrink: 0;
  width: 122px;
  margin-right: 16px;
}
