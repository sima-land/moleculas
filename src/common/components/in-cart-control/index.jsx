import React from 'react';
import { Button } from '@dev-dep/ui-nucleons/button';
import { Stepper } from '@dev-dep/ui-nucleons/stepper';
import { useCartItemInput } from './utils';
import SpinnerSVG from '../../../common/icons/white-spinner.svg';
import classnames from 'classnames/bind';
import styles from './in-cart-control.scss';

const cx = classnames.bind(styles);

/**
 * Компонент управления количеством в корзине.
 * @param {Object} props Свойства.
 * @return {ReactElement} Элемент.
 */
export const InCartControl = ({
  canAdd,
  canSubtract,
  className,
  onAdd,
  onChange,
  onSubtract,
  qty,
  isFetching,
  allowFloat,
  inCart,
}) => {
  const inputProps = useCartItemInput({ qty, allowFloat, onApply: onChange });

  return (
    <>
      {
        inCart
          ? (
            <Stepper
              size='s'
              disabled={isFetching}
              className={cx('control', className)}
              canAdd={canAdd}
              canSubtract={canSubtract}
              onAdd={() => onAdd && onAdd()}
              onSubtract={() => onSubtract && onSubtract()}
              {...inputProps}
            />
          ) : (
            <Button
              size='small'
              className={cx('control', className)}
              onClick={() => !isFetching && onAdd && onAdd()}
            >
              {isFetching
                ? (
                  <SpinnerSVG
                    className={cx('button-spinner')}
                    width={24}
                    height={24}
                  />
                ) : 'В корзину'
              }
            </Button>
          )
      }
    </>
  );
};
