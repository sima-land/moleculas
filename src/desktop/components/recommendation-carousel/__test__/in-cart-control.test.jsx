import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { InCartControl } from '../in-cart-control';
import { Stepper } from '@dev-dep/ui-nucleons/stepper';

describe('<InCartControl />', () => {
  it('should renders correctly', () => {
    const Spy = {
      onAdd: jest.fn(),
      onChange: jest.fn(),
      onSubtract: jest.fn(),
    };

    const wrapper = mount(
      <InCartControl
        qty={12}
        canAdd
        canSubtract
        markupText='foo'
        stepText='bar'
        onAdd={Spy.onAdd}
        onChange={Spy.onChange}
        onSubtract={Spy.onSubtract}
      />
    );

    expect(wrapper).toMatchSnapshot();

    Object.values(Spy).forEach(fn => {
      expect(fn).toBeCalledTimes(0);
    });

    // add
    act(() => {
      wrapper.find(Stepper).find('svg[data-testid="stepper:plus"]').simulate('click');
    });
    wrapper.update();

    expect(Spy.onAdd).toBeCalledTimes(1);

    // subtract
    act(() => {
      wrapper.find(Stepper).find('svg[data-testid="stepper:minus"]').simulate('click');
    });
    wrapper.update();

    expect(Spy.onSubtract).toBeCalledTimes(1);

    // subtract
    act(() => {
      wrapper.find(Stepper).find('[data-testid="stepper:input"]').simulate('blur', {
        target: {
          value: 15,
        },
      });
    });
    wrapper.update();

    // enter keydown
    act(() => {
      wrapper.find(Stepper).find('[data-testid="stepper:input"]').simulate('keydown', {
        key: 'Enter',
      });
    });
    wrapper.update();

    expect(Spy.onChange).toBeCalledTimes(2);
  });
});
