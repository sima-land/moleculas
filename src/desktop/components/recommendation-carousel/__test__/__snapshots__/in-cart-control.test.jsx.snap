// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InCartControl /> should renders correctly 1`] = `
<InCartControl
  canAdd={true}
  canSubtract={true}
  markupText="foo"
  onAdd={[MockFunction]}
  onChange={[MockFunction]}
  onSubtract={[MockFunction]}
  qty={12}
  stepText="bar"
>
  <Stepper
    canAdd={true}
    canSubtract={true}
    className="control"
    onAdd={[MockFunction]}
    onBlur={[MockFunction]}
    onChange={[Function]}
    onKeyDown={[Function]}
    onSubtract={[MockFunction]}
    size="s"
    value={12}
  >
    <div
      className="root size-s control"
    >
      <MinusSVG
        className="button"
        data-testid="stepper:minus"
        onClick={[MockFunction]}
      >
        <svg
          className="button"
          data-testid="stepper:minus"
          height={16}
          onClick={[MockFunction]}
          viewBox="0 0 16 16"
          width={16}
        >
          <path
            clipRule="evenodd"
            d="M2 8a1 1 0 011-1h10a1 1 0 110 2H3a1 1 0 01-1-1z"
            fillRule="evenodd"
          />
        </svg>
      </MinusSVG>
      <input
        className="input"
        data-testid="stepper:input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        value={12}
      />
      <PlusSVG
        className="button"
        data-testid="stepper:plus"
        onClick={[MockFunction]}
      >
        <svg
          className="button"
          data-testid="stepper:plus"
          height={16}
          onClick={[MockFunction]}
          viewBox="0 0 16 16"
          width={16}
        >
          <path
            d="M9 3a1 1 0 00-2 0v4H3a1 1 0 000 2h4v4a1 1 0 102 0V9h4a1 1 0 100-2H9V3z"
          />
        </svg>
      </PlusSVG>
    </div>
  </Stepper>
</InCartControl>
`;
